### PR TITLE
Removed duplicate Koenig node renderers from Ghost core

### PR DESCRIPTION
**Pull request by kevinansfield**
ref https://linear.app/ghost/issue/BER-3775/

Now that Koenig lives in the Ghost monorepo, the forked node renderers in `ghost/core/core/server/services/koenig/` no longer serve a purpose — the fork only existed to avoid the cross-repo PR/publish/bump dance when changing render output, and the two renderer sets had drifted apart since the fork was created in Feb 2025.

This makes `kg-default-nodes` the single source of truth for card render output: the Ghost fork (which had received all production changes since it was created) was synced back into the `kg-default-nodes` renderers and then deleted, so render output changes are now made in one place with no risk of the copies drifting again.

- all 24 renderers plus the shared render helpers were synced; stale unshipped divergences on the Koenig side (old `emailCustomization`/`emailCustomizationAlpha` flag branches, pre-fork vintages of the outline-button and button-corners handling) were replaced with the shipped Ghost logic
- output parity was verified byte-for-byte by rendering the golden-post kitchen-sink fixture through the real render pipeline with both renderer sets, across web/email targets with all feature flags and design options enabled
- Ghost's per-renderer snapshot test suites moved into `kg-default-nodes` alongside the renderers they cover, and the existing Koenig node tests were updated where their expectations encoded the stale pre-sync output
- the `options.nodeRenderers` override mechanism remains supported in `kg-default-nodes` for external consumers, but Ghost no longer injects a custom map; the default renderers are load-bearing regardless as Lexical calls `exportDOM` for clipboard HTML on cut/copy (see the revert in `bdc08e5e89`)
- image `<picture>`/avif/webp support is now gated on a `canTransformImageToFormat` render option injected by Ghost's `lexical.js` instead of importing the server-only `@tryghost/image-transform` inside `kg-default-nodes`, which is bundled into the editor's browser build

A handful of intentional behaviour fixes to edge-case output are included (each covered by a renderer test):

- the fork's header v1 renderer called `.replace(/(<br>)+$/g)` with no replacement argument, so a subheading consisting only of trailing `<br>` tags rendered as the literal string "undefined" — the corrected Koenig implementation is kept
- CTA card colour validation was unanchored, so invalid `buttonColor`/`backgroundColor` values could be interpolated into `style`/`class` attributes instead of falling back to defaults
- video cards and video embeds rendered `NaN`/`Infinity` spacer dimensions in emails when width/height metadata was missing or non-numeric
- twitter embeds with incomplete `tweet_data` emitted `status/undefined` links, "Invalid DateTime" timestamps, and image rows with no image — they now fall back gracefully
- audio/video/file cards rendered to email without a `postUrl` (e.g. member welcome emails) threw or emitted empty links; they now fall back to linking the media file
- button cards rendered an empty link in email when `buttonText` was blank; they now use the same "Button Title" fallback as the web output
- image email rendering crashed when image transforms were enabled without `imageOptimization` config

### CODE REVIEW

File: koenig/kg-default-nodes/src/nodes/video/video-renderer.ts [hunk 1]
```diff
@@ -137,13 +128,13 @@ export function cardTemplate({node, cardClasses}: {node: VideoNodeData, cardClas
     );
 }
 
-export function emailCardTemplate({node, options, cardClasses}: {node: VideoNodeData, options: EmailVideoRenderOptions, cardClasses: string}) {
+function emailCardTemplate({node, options, cardClasses}: {node: VideoNodeData, options: EmailVideoRenderOptions, cardClasses: string}) {
     const thumbnailSrc = node.customThumbnailSrc || node.thumbnailSrc;
     const emailTemplateMaxWidth = 600;
-    const aspectRatio = hasVideoDimensions(node) ? node.width / node.height : DEFAULT_EMAIL_ASPECT_RATIO;
+    const aspectRatio = node.width! / node.height!;
```
[the remark this answers is not in the document]
Comment (kevinansfield) [reply koenig/kg-default-nodes/src/nodes/video/video-renderer.ts]: Fixed in 0b7690178e — email rendering now falls back to a 16:9 aspect ratio when width/height are missing, so no NaN spacer URLs/styles.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: koenig/kg-default-nodes/src/nodes/image/image-renderer.ts [hunk 2]
```diff
@@ -116,31 +183,27 @@ export function renderImageNode(node: ImageNodeData, options: ImageRenderOptions
         img.setAttribute('width', String(imageDimensions.width));
         img.setAttribute('height', String(imageDimensions.height));
 
-        const contentImageSizes = options.imageOptimization?.contentImageSizes;
-        if (contentImageSizes && isLocalContentImage(node.src, options.siteUrl) && options.canTransformImage?.(node.src)) {
+        if (isContentImage(node.src, options.siteUrl, options.imageBaseUrl) && options.canTransformImage?.(node.src)) {
             // find available image size next up from 2x600 so we can use it for the "retina" src
-            const availableImageWidths = getAvailableImageWidths(node, contentImageSizes);
+            const availableImageWidths = getAvailableImageWidths(node, options.imageOptimization!.contentImageSizes!);
```
[the remark this answers is not in the document]
Comment (kevinansfield) [reply koenig/kg-default-nodes/src/nodes/image/image-renderer.ts]: Fixed in 0b7690178e — the retina rewrite is now skipped unless imageOptimization.contentImageSizes exists, and the content-images path match is guarded.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: ghost/core/core/server/lib/lexical.js [hunk 3]
```diff
@@ -94,25 +96,8 @@ module.exports = {
...
-                });
-            }
-        }
-
-        return customNodeRenderers;
-    },
-
     async render(lexical, userOptions = {}) {
-        const options = buildRenderOptions(userOptions, this.customNodeRenderers);
+        const options = buildRenderOptions(userOptions);
```
[the remark this answers is not in the document]
Comment (kevinansfield) [reply ghost/core/core/server/lib/lexical.js:100]: Fixed in 0b7690178e on the renderer side — audio/video (and file) cards no longer throw without postUrl; they fall back to linking the media file, so welcome emails render fine.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: koenig/kg-default-nodes/src/nodes/embed/embed-renderer.ts [hunk 4]
```diff
@@ -44,9 +44,9 @@ function renderTemplate(node: EmbedNodeData, document: Document, options: Render
     const figure = document.createElement('figure');
     figure.setAttribute('class', 'kg-card kg-embed-card');
 
-    if (isEmail && isVideoWithThumbnail && metadata.thumbnail_width && metadata.thumbnail_height) {
+    if (isEmail && isVideoWithThumbnail) {
         const emailTemplateMaxWidth = 600;
-        const thumbnailAspectRatio = metadata.thumbnail_width / metadata.thumbnail_height;
+        const thumbnailAspectRatio = Number(metadata.thumbnail_width) / Number(metadata.thumbnail_height);
```
[the remark this answers is not in the document]
Comment (kevinansfield) [reply koenig/kg-default-nodes/src/nodes/embed/embed-renderer.ts]: Fixed in 0b7690178e — the email video preview path now requires thumbnail_width/height and falls back to the standard figure otherwise.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: koenig/kg-default-nodes/src/nodes/call-to-action/calltoaction-renderer.ts [hunk 5]
```diff
@@ -46,7 +50,7 @@ const wrapWithLink = (dataset: CTADataset, content: string) => {
 
 function ctaCardTemplate(dataset: CTADataset) {
     // Add validation for buttonColor
-    if (!dataset.buttonColor || !dataset.buttonColor.match(/^(?:[a-zA-Z\d-]+|#(?:[a-fA-F\d]{3}|[a-fA-F\d]{6}))$/)) {
+    if (!dataset.buttonColor || !dataset.buttonColor.match(/^[a-zA-Z\d-]+|#([a-fA-F\d]{3}|[a-fA-F\d]{6})$/)) {
```
[the remark this answers is not in the document]
Comment (kevinansfield) [reply koenig/kg-default-nodes/src/nodes/call-to-action/calltoaction-renderer.ts]: Fixed in 0b7690178e — restored the fully anchored validation for buttonColor (and backgroundColor, which had the same gap).

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: koenig/kg-default-nodes/src/nodes/embed/types/twitter.ts [hunk 6]
```diff
@@ -68,7 +68,7 @@ export default function render(node: TwitterNode, document: Document, options: R
     const tweetData = metadata && metadata.tweet_data;
     const isEmail = options.target === 'email';
 
-    if (tweetData && isEmail && tweetData.id) {
+    if (tweetData && isEmail) {
         const tweetId = tweetData.id;
```
[the remark this answers is not in the document]
Comment (kevinansfield) [reply koenig/kg-default-nodes/src/nodes/embed/types/twitter.ts]: Fixed in 0b7690178e — the email tweet card only renders when tweet_data.id is present, otherwise the original embed html is used.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: koenig/kg-default-nodes/src/nodes/button/button-renderer.ts [hunk 7]
```diff
@@ -42,69 +42,30 @@ function frontendTemplate(node: ButtonNodeData, document: Document) {
...
-        `;
-
-        const element = document.createElement('p');
-        element.innerHTML = cardHtml;
-        return {element, type: 'outer' as const};
-    }
+    const element = document.createElement('div');
+    element.innerHTML = cardHtml;
+    return {element, type: 'inner' as const};
 }
```
[the remark this answers is not in the document]
Comment (kevinansfield) [reply koenig/kg-default-nodes/src/nodes/button/button-renderer.ts:69]: Fixed in 0b7690178e — the email template now falls back to "Button Title", matching the frontend template.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: koenig/kg-default-nodes/src/nodes/call-to-action/calltoaction-renderer.ts [hunk 8]
```diff
@@ -46,7 +50,7 @@ const wrapWithLink = (dataset: CTADataset, content: string) => {
...
         dataset.buttonColor = 'accent';
     }
```
[the remark this answers is not in the document]
Comment (kevinansfield) [reply koenig/kg-default-nodes/src/nodes/call-to-action/calltoaction-renderer.ts:55]: Fixed in 0b7690178e — anchored the regex for both buttonColor and backgroundColor.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: koenig/kg-default-nodes/src/nodes/embed/types/twitter.ts [hunk 9]
```diff
@@ -79,9 +79,8 @@ export default function render(node: TwitterNode, document: Document, options: R
         const retweetCount = numberFormatter.format(tweetData.public_metrics?.retweet_count ?? 0);
         const likeCount = numberFormatter.format(tweetData.public_metrics?.like_count ?? 0);
         const authorUser = tweetData.users && tweetData.users.find((user: TwitterUser) => user.id === tweetData.author_id);
-        const parsedCreatedAt = tweetData.created_at ? DateTime.fromISO(tweetData.created_at) : null;
-        const tweetTime = parsedCreatedAt?.isValid ? parsedCreatedAt.toLocaleString(DateTime.TIME_SIMPLE) : '';
-        const tweetDate = parsedCreatedAt?.isValid ? parsedCreatedAt.toLocaleString(DateTime.DATE_MED) : '';
+        const tweetTime = DateTime.fromISO(tweetData.created_at ?? '').toLocaleString(DateTime.TIME_SIMPLE);
+        const tweetDate = DateTime.fromISO(tweetData.created_at ?? '').toLocaleString(DateTime.DATE_MED);
```
[the remark this answers is not in the document]
Comment (kevinansfield) [reply koenig/kg-default-nodes/src/nodes/embed/types/twitter.ts]: Fixed in 0b7690178e — the timestamp is only rendered when the parsed DateTime is valid.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: koenig/kg-default-nodes/src/nodes/embed/types/twitter.ts [hunk 10]
```diff
@@ -90,12 +89,11 @@ export default function render(node: TwitterNode, document: Document, options: R
...
         let tweetImageUrl = null;
-        const firstMedia = tweetData.includes?.media?.[0];
-        const hasImageOrVideo = tweetData.attachments?.media_keys && firstMedia;
+        const hasImageOrVideo = tweetData.attachments && tweetData.attachments.media_keys;
         if (hasImageOrVideo) {
-            tweetImageUrl = firstMedia.preview_image_url || firstMedia.url;
+            tweetImageUrl = tweetData.includes?.media?.[0]?.preview_image_url || tweetData.includes?.media?.[0]?.url;
         }
-        const hasPoll = tweetData.attachments && tweetData.attachments && tweetData.attachments.poll_ids;
+        const hasPoll = tweetData.attachments && tweetData.attachments.poll_ids;
```
[the remark this answers is not in the document]
Comment (kevinansfield) [reply koenig/kg-default-nodes/src/nodes/embed/types/twitter.ts]: Fixed in 0b7690178e — the image row is now gated on a resolved media URL from includes.media, not just media_keys.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: koenig/kg-default-nodes/src/nodes/file/file-renderer.ts [hunk 11]
```diff
@@ -58,24 +47,22 @@ function emailTemplate(node: FileNodeData, document: Document, options: RenderOp
...
                                 <table cellspacing="0" cellpadding="0" border="0" width="100%"><tr><td>
-                                    ${wrapWithAnchor(`<span class="kg-file-name">${escapeHtml(node.fileName)}</span> &bull; ${bytesToSize(node.fileSize)}`, href, 'kg-file-meta')}
+                                    <a href="${escapeHtml(options.postUrl)}" class="kg-file-meta"><span class="kg-file-name">${escapeHtml(node.fileName)}</span> &bull; ${bytesToSize(node.fileSize)}</a>
                                 </td></tr></table>
                             </td>
                             <td width="80" valign="middle" class="kg-file-thumbnail">
-                                ${href
-            ? `<a href="${escapeHtml(href)}" style="display: block; top: 0; right: 0; bottom: 0; left: 0;">
+                                <a href="${escapeHtml(options.postUrl)}" style="display: block; top: 0; right: 0; bottom: 0; left: 0;">
                                     <img src="https://static.ghost.org/v4.0.0/images/download-icon-darkmode.png" style="${escapeHtml(iconCls)}">
```
[the remark this answers is not in the document]
Comment (kevinansfield) [reply koenig/kg-default-nodes/src/nodes/file/file-renderer.ts]: Fixed in 0b7690178e — the email links fall back to node.src when options.postUrl is absent.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: koenig/kg-default-nodes/src/nodes/header/renderers/v2/header-renderer.ts [hunk 12]
```diff
@@ -117,97 +118,129 @@ function cardTemplate(nodeData: HeaderV2NodeData, options: HeaderV2RenderOptions
...
-            `
-        );
-    }
+    const buttonHtml = renderEmailButton({
+        url: nodeData.buttonUrl,
+        text: nodeData.buttonText,
+        alignment: nodeData.alignment,
+        color: nodeData.buttonColor,
+        style: options?.design?.buttonStyle
+    });
```
[the remark this answers is not in the document]
Comment (kevinansfield) [reply koenig/kg-default-nodes/src/nodes/header/renderers/v2/header-renderer.ts:202]: Left as-is: renderEmailButton special-cases "accent" — it is excluded from colored fill/outline styling and mapped to the btn-accent class, which the email template CSS styles with the site accent color. Resolving to a hex here would bypass that and diverge from the frontend path.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: koenig/kg-default-nodes/src/nodes/image/image-renderer.ts [hunk 13]
```diff
@@ -116,31 +183,27 @@ export function renderImageNode(node: ImageNodeData, options: ImageRenderOptions
...
             if (!srcWidth || srcWidth === node.width) {
                 // do nothing, width is smaller than retina or matches the original node src
             } else {
-                const match = node.src.match(/(.*\/content\/images)\/(.*)/);
-                if (match) {
-                    const [, imagesPath, filename] = match;
-                    img.setAttribute('src', `${imagesPath}/size/w${srcWidth}/${filename}`);
-                }
+                const [, imagesPath, filename] = node.src.match(/(.*\/content\/images)\/(.*)/)!;
+                img.setAttribute('src', `${imagesPath}/size/w${srcWidth}/${filename}`);
```
[the remark this answers is not in the document]
Comment (kevinansfield) [reply koenig/kg-default-nodes/src/nodes/image/image-renderer.ts]: Fixed in 0b7690178e — guarded on imageOptimization?.contentImageSizes and on the content-images path match before rewriting the src.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: koenig/kg-default-nodes/src/nodes/transistor/transistor-renderer.ts [hunk 14]
```diff
@@ -1,52 +1,200 @@
...
+    const transistorUrl = 'https://partner.transistor.fm/ghost/{uuid}';
+
+    const cardHtml = html`
+        <table class="kg-card kg-transistor-card" cellspacing="0" cellpadding="0" border="0" width="100%">
+            <tr>
+                <td style="padding: 24px; text-align: center;">
+                    <table cellspacing="0" cellpadding="0" border="0" width="100%" style="text-align: center;">
+                        <tr>
+                            <td style="text-align: center; padding-bottom: 12px;">
+                                <a href="${transistorUrl}" style="display: inline-block; width: 72px; height: 72px; padding-top: 4px; padding-right: 4px; padding-bottom: 4px; padding-left: 4px; border-radius: 8px; background-color: ${accentColor}">
```
[the remark this answers is not in the document]
Comment (kevinansfield) [reply koenig/kg-default-nodes/src/nodes/transistor/transistor-renderer.ts:83]: Left as-is: options.design.accentColor is the site's own design setting (owner-controlled via admin), not per-card user data, so validating it here adds no security boundary. It is also interpolated the same way by the other email card templates.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: koenig/kg-default-nodes/src/utils/tagged-template-fns.ts [hunk 15]
```diff
@@ -1,15 +1,23 @@
...
 
     // Handle tagged template literal case
     const result = strings.reduce((acc: string, str: string, i: number) => {
-        return acc + str + (values[i] ?? '');
+        return acc + str + (values[i] || '');
     }, '');
     // Remove newline+indentation patterns while preserving intentional whitespace
-    return result.replace(/\n\s+/g, '').trim();
+    return _oneline(result);
 };
```
[the remark this answers is not in the document]
Comment (kevinansfield) [reply koenig/kg-default-nodes/src/utils/tagged-template-fns.ts:21]: Left as-is deliberately: the tagged templates rely on falsy interpolations collapsing to empty strings (e.g. `${templateData.hasHeader && html`...`}` in the header renderer) — switching to `??` would render literal "false" into output. No renderer legitimately interpolates 0/false/NaN as content.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: koenig/kg-default-nodes/test/renderers/call-to-action-renderer.test.ts [hunk 16]
```diff
@@ -137,6 +226,119 @@
...
+                    internalImageSizes: {
+                        'email-cta-minimal-image': {width: 64, height: 64}
+                    }
+                }
+            });
+
+            assert.ok(result.html.includes('kg-cta-bg-green'));
+            assert.ok(result.html.includes('background-color: #F0F0F0'));
+            assert.ok(result.html.includes('Get access now'));
+            assert.ok(result.html.includes('http://someblog.com/somepost'));
```
[the remark this answers is not in the document]
Comment (kevinansfield) [reply koenig/kg-default-nodes/test/renderers/call-to-action-renderer.test.ts]: Fixed in 0b7690178e — the assertion now matches the full href attribute rather than a URL substring.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: koenig/kg-default-nodes/test/renderers/embed-renderer.test.ts [hunk 17]
```diff
@@ -152,5 +216,75 @@
...
+                    tweet_data: {
+                        text: 'tweet text without an id'
+                    },
+                    type: 'rich',
+                    version: '1.0'
+                }
+            }));
+
+            assert.ok(result.html.includes('<table cellspacing="0" cellpadding="0" border="0" class="kg-twitter-card">'));
+            assert.ok(result.html.includes('https://twitter.com/twitter/status/undefined'));
```
[the remark this answers is not in the document]
Comment (kevinansfield) [reply koenig/kg-default-nodes/test/renderers/embed-renderer.test.ts]: Fixed in 0b7690178e — with missing tweet_data.id the renderer now falls back to the original embed html, and the test asserts no status/undefined link is produced.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
